kvcoord: only add to the buffer on the response path

arulajmani opened the pull request:
See individual commits for details.

arulajmani committed:
kvclient: add tests for CPut evaluation with buffered writes

ConditionFailedErrors returned by CPuts are special, in that they're
non-terminal. That means that a transaction is allowed to commit even
if it gets one of these.

This patch reverts changes to tests from da7a4eb, which assumed we won't
try to commit transactions that encountered ConditionFailedErrors. While here,
we also extend TestTxnContinueAfterCputError to run with buffered writes and
validate the transaction's effects upon commit.

All these tests fail as of this commit.

Epic: none

Release note: None

arulajmani committed:
kvcoord: only add to the buffer on the response path

This patch moves all additions to the write buffer on the response
path of a request. This cleans things up for CPuts, where in the
old structure, we were forced to optimistically add writes to the
buffer so that subsequent requests in the batch saw them. This posed
an issue whenver there was a ConditionFailedError, as a transaction
is allowed to commit despite this.

Closes https://github.com/cockroachdb/cockroach/issues/139055
Release note: None

arulajmani committed:
kvclient: serve reads from the buffer on the response path

Previously, we were serving reads from the buffer when transforming
a batch. Because of the last commit, this meant that any writes from
earlier in the batch that needed to be added to the buffer would be
added at response time, so the reads served on the request path are
no longer correct. This patch fixes the issue by serving reads from
the buffer on the response path.

Currently, we're not being smart about omitting Get requests for keys
that are reading keys that were written to earlier in the batch. We
could change this behaviour in the future if we think its worthwhile.
SQL never constructs batches that read and write to the same key, so
this is likely not a big deal for now.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144584)
<!-- Reviewable:end -->

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Just some minor clarifications.

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-:-OO4Nt63Dq7W4bgvSyUr:bbl5nnx)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @stevendanna and @yuzefovich)

___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4-IYQ8ZWsZ-ejgfBw:-OO4-IYQ8ZWsZ-ejgfBx:b1y9zgg):*
Is this ok? Wouldn't it mess up running tests on this commit (e.g. as part of bisect)?
___
*[`-- commits` line 36 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4NUIp5faU5PIs9zbe:-OO4EAOr8z8CgFZ98PkP:bov56ny):*
Shouldn't this sentence be more like: "Because of the last commit, this meant that any writes from earlier in the batch that needed to be added to the buffer would be added at response time, so the read served at transform time may no longer be correct."?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 464 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4BX_I3KXAslQ_dhAd:-OO4BX_I3KXAslQ_dhAe:bgv2zj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118a3686f8d8af610fc4548a12ed9cf09b4d5a8/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L464)):*
> ```Go
> 			require.Equal(t, resErrIdx, pErr.Index.Index)
> 
> 			// The batch we sent encountered an error; nothing should have been
> ```

This seems like a nicer behavior than before!
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 448 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4Lmeo0m_0G95XSNHA:-OO4Lmeo0m_0G95XSNHB:b-702p8a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a857c4bba1c4e7c0dc01abfefca40638b42333af/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L448)):*
> ```Go
> 					}
> 				}
> 				if curTs.origRequest.Method() != kvpb.Get {
> ```

Why do we do this?
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1096 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO48RntD12AYk9B-zQt:-OO48RntD12AYk9B-zQu:bnap07p) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118a3686f8d8af610fc4548a12ed9cf09b4d5a8/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1096)):*
> ```Go
> 		)
> 		if condFailedErr != nil {
> 			// TODO(yuzefovich): consider "poisoning" the txnWriteBuffer when we
> ```

Reminder to remove this as we discussed.


<!-- Sent from Reviewable.io -->

miraradeva commented:
The CI failure seems legit though.

miraradeva reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144584)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @yuzefovich)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4RXdUAoZs2Z03mWQH:-OO4RXdUAoZs2Z03mWQI:bw43tya) ([raw file](https://github.com/cockroachdb/cockroach/blob/a857c4bba1c4e7c0dc01abfefca40638b42333af/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1115)):*
> ```Go
> 
> 	case *kvpb.DeleteRequest:
> 		twb.addToBuffer(req.Key, roachpb.Value{}, req.Sequence)
> ```

This is probably not quite right: if we add the delete to the buffer here, `maybeServeRead` below will see it, with an empty value, and will set `foundKey = false` regardless of whether we actually deleted anything.


<!-- Sent from Reviewable.io -->

arulajmani committed:
kvcoord: test that FoundKey is correctly populated on DeleteResponses

If there were a bug here, our logic tests would catch this, as evidenced
by the CI failure on
https://github.com/cockroachdb/cockroach/pull/144584. Nevertheless, add
a more direct test.

Epic: none

Release note: None

arulajmani reviewed (commented):
TFTR! Like we spoke about offline, I added a more direct test for the CI failure, and it does fail before the fix. RFAL. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-:-OO8kmwQ8qjC_KSHHrhA:bqvupt8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @miraradeva, @stevendanna, and @yuzefovich)

___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4-IYQ8ZWsZ-ejgfBw:-OO8BHq59NNdS5PhQndn:b-af800y):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Is this ok? Wouldn't it mess up running tests on this commit (e.g. as part of bisect)?
</blockquote></details>

Just for these tests, so I think it's fine. 
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1096 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO48RntD12AYk9B-zQt:-OO8bjfUAV3DzpVgbhf8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118a3686f8d8af610fc4548a12ed9cf09b4d5a8/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1096)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Reminder to remove this as we discussed.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 448 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4Lmeo0m_0G95XSNHA:-OO8IuwI9475y7qQPDps:b-r6zjgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a857c4bba1c4e7c0dc01abfefca40638b42333af/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L448)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Why do we do this?
</blockquote></details>

Mostly to keep the current behaviour intact. We weren't handling this else case correctly to begin with, which prompted https://github.com/cockroachdb/cockroach/issues/144694. I imagine this will change a fair bit once we address that issue and improve testing.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 1115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4RXdUAoZs2Z03mWQH:-OO8barO1nes880rvgEn:b-xspzk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a857c4bba1c4e7c0dc01abfefca40638b42333af/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L1115)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

This is probably not quite right: if we add the delete to the buffer here, `maybeServeRead` below will see it, with an empty value, and will set `foundKey = false` regardless of whether we actually deleted anything.
</blockquote></details>

Yeah, you're right -- that's what caused the CI failures; fixed.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go` line 464 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-OO4BX_I3KXAslQ_dhAd:-OO8IfYwC3Bb1Ljf3wqo:b-pbdqzg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b118a3686f8d8af610fc4548a12ed9cf09b4d5a8/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go#L464)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

This seems like a nicer behavior than before!
</blockquote></details>

Agreed!


<!-- Sent from Reviewable.io -->

miraradeva reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r4, 2 of 2 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-:-OO8oH1669BFpOQaxYha:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @yuzefovich)



<!-- Sent from Reviewable.io -->

arulajmani commented:
Thanks Mira!

bors r+

arulajmani commented:
blathers backport 25.2

yuzefovich reviewed (commented):
Nice, thanks!

> Reviewed 3 of 3 files at r4, 2 of 2 files at r5, 2 of 2 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144584#-:-OODCyb17BhQeZldkop6:b-wdj8zi)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->
